smartsim/entity/application.py

```diff
@@ -0,0 +1,291 @@
+
+        :param to_copy: files to copy
+        :param to_symlink: files to symlink
+        :param to_configure: input files with tagged parameters
+        :raises ValueError: if the generator file already exists
+        """
+        to_copy = to_copy or []
+        to_symlink = to_symlink or []
+        to_configure = to_configure or []
+
+        # Check that no file collides with the parameter file written
+        # by Generator. We check the basename, even though it is more
+        # restrictive than what we need (but it avoids relative path issues)
+        for strategy in [to_copy, to_symlink, to_configure]:
+            if strategy is not None and any(
+                osp.basename(filename) == "smartsim_params.txt" for filename in strategy
+            ):
+                raise ValueError(
+                    "`smartsim_params.txt` is a file automatically "
+                    + "generated by SmartSim and cannot be ovewritten."
+                )
+        self.files = EntityFiles(to_configure, to_copy, to_symlink)
+
+    @property
+    def attached_files_table(self) -> str:
+        """Return a list of attached files as a plain text table
+
+        :returns: String version of table
+        """
+        if not self.files:
+            return "No file attached to this application."
+        return str(self.files)
+
+    def print_attached_files(self) -> None:
+        """Print a table of the attached files on std out"""
+        print(self.attached_files_table)
+
+    def __str__(self) -> str:  # pragma: no cover
+        exe_args_str = "\n".join(self.exe_args)
+        entities_str = "\n".join(str(entity) for entity in self.incoming_entities)
+        return textwrap.dedent(f"""\
+            Name: {self.name}
+            Type: {self.type}
+            Executable:
+            {self.exe}
+            Executable Arguments:
+            {exe_args_str}
+            Entity Files: {self.files}
+            File Parameters: {self.file_parameters}
+            Incoming Entities:
+            {entities_str}
+            Key Prefixing Enabled: {self.key_prefixing_enabled}
+            """)
+
+    @staticmethod
+    def _build_exe_args(exe_args: t.Union[str, t.Sequence[str], None]) -> t.List[str]:
+        """Check and convert exe_args input to a desired collection format
+
+        :param exe_args:
+        :raises TypeError: if exe_args is not a list of str or str
+        """
+        if not exe_args:
+            return []
+
+        if isinstance(exe_args, list):
+            exe_args = copy.deepcopy(exe_args)
+
+        if not (
+            isinstance(exe_args, str)
+            or (
+                isinstance(exe_args, list)
+                and all(isinstance(arg, str) for arg in exe_args)
+            )
+        ):
+            raise TypeError("Executable arguments were not a list of str or a str.")
+
+        if isinstance(exe_args, str):
+            return exe_args.split()
+
+        return exe_args
```

Comment:
One small thing: I thought we decided to remove the extra check here and deep copy at the return point:
```suggestion
        if not (
            isinstance(exe_args, str)
            or (
                isinstance(exe_args, list)
                and all(isinstance(arg, str) for arg in exe_args)
            )
        ):
            raise TypeError("Executable arguments were not a list of str or a str.")

        if isinstance(exe_args, str):
            return exe_args.split()

        return copy.deepcopy(exe_args)
```

Reply:
you right, wonky merge

Reply:
changed